### PR TITLE
WIP feat: paylod validation

### PR DESCRIPTION
**Pull request by Khaledgarbaya**
- [x] Add more tests
- [x] Improve error message?

### CODE REVIEW

**Comment by Khaledgarbaya** [comment]
aaaaaand `Joi` does not support node `v4.7` 🤦‍♂️

File: lib/utils/schema.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  fields: Joi.object()
+}
+
+const contentTypeSchema = {
+  sys: Joi.object(),
+  fields: Joi.array().required().items(Joi.object().keys({
+    id: Joi.string().required(),
+    name: Joi.string().required(),
+    type: Joi.string().required().regex(/Symbol|Text|Integer|Number|Date|Object|Boolean|Array|Link/),
+    validations: Joi.array(),
```
Comment (axe312ger) [inline lib/utils/schema.js]: Shouldn't we have a schema for validations?

Comment (Khaledgarbaya) [reply lib/utils/schema.js]: validations is so vague and it depends on the field. We can live without validating the validations for now, also we didn't have any issue related to the validations

Comment (axe312ger) [reply lib/utils/schema.js]: Alright, sounds good to me

**Comment by Khaledgarbaya** [comment]
@axe312ger this PR is ready for review, `Joi` by default was stopping early, that's why it only shows one error. I disabled that and the error message telling you how many entities are invalid.

**Comment by axe312ger** [comment]
Neat, seeing all errors at once is amazing 🎉 

But the output in the CLI still could use some readability:
```
15:05:85 - ValidationError: Payload validation error: Invalid assets:7 Invalid locales:1 Invalid webhooks:1 Get further details in the error log file
```

How about:
```
15:05:85 - ValidationError: 7 assets, 1 locales, 1 webhooks - Get further details in the error log file
```
(Maybe even check if it is multiple and append `s` to make it plural)

File: lib/utils/validations.js [hunk 2]
```diff
@@ -19,17 +19,15 @@ const countInvalidEntities = (validationData) => {
...
   }, {})
-  return Object.keys(entityCount).map((key) => `Invalid ${key}:${entityCount[key]}`)
+  return Object.keys(entityCount).map((key) => `${key}:${entityCount[key]}`)
 }
 
 const assertPayload = (payload) => {
   const result = Joi.validate(payload, payloadSchema, {allowUnknown: true, abortEarly: false})
   if (result.error) {
     attachEntityName(result.error.details, payload)
     const invalidEntityCount = countInvalidEntities(result.error.details).join('\n')
```
Comment (axe312ger) [inline lib/utils/validations.js]: I'd join by `, ` not by newline. We have only one line available here.

File: errorlogfile [hunk 3]
```diff
@@ -0,0 +1,106 @@
+[
```
Comment (axe312ger) [inline errorlogfile]: Is this file needed? I don't think so 🤷‍♂️

File: lib/utils/validations.js [hunk 4]
```diff
@@ -1,3 +1,38 @@
...
+    return entities
+  }, {})
+  return Object.keys(entityCount).map((key) => `${key}:${entityCount[key]}`)
+}
+
+const assertPayload = (payload) => {
+  const result = Joi.validate(payload, payloadSchema, {allowUnknown: true, abortEarly: false})
+  if (result.error) {
+    attachEntityName(result.error.details, payload)
+    const invalidEntityCount = countInvalidEntities(result.error.details).join('\n')
```
Comment (axe312ger) [inline lib/utils/validations.js]: Lets join via comma + space instead of new line since we have only one line available in the Listr task error display.

File: test/unit/validations.test.js [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+      contentType: 'image/png'
+    }
+  }
+  expect(() => assertPayload({
+    entries: [],
+    locales: [],
+    contentTypes: [],
+    assets: [brokenAsset, brokenAsset]
+  })
+  ).toThrowErrorMatchingSnapshot()
```
Comment (axe312ger) [inline test/unit/validations.test.js]: Jest ❤️

File: lib/utils/validations.js [hunk 6]
```diff
@@ -1,3 +1,38 @@
...
+const countInvalidEntities = (validationData) => {
+  const entityCount = validationData.reduce((entities, currentDetail) => {
+    if (!entities[currentDetail.path[0]]) {
+      entities[currentDetail.path[0]] = 1
+    } else {
+      entities[currentDetail.path[0]]++
+    }
+    return entities
+  }, {})
+  return Object.keys(entityCount).map((key) => `${key}:${entityCount[key]}`)
```
Comment (axe312ger) [inline lib/utils/validations.js]: What about if we pass key through https://lodash.com/docs/4.17.4#startCase. IMHO that would improve readability :)

(We probably could do this then later on on more parts in the code to properly format entity types)

File: test/unit/index.test.js [hunk 7]
```diff
@@ -5,10 +5,15 @@ import pushToSpaceStub from 'contentful-batch-libs/dist/push/push-to-space'
...
 import createClientsStub from 'contentful-batch-libs/dist/utils/create-clients'
 import getDestinationResponseStub from '../../lib/get-destination-response'
-
 import contentfulImport from '../../lib/index'
 
 jest.mock('cli-table2')
+jest.mock('../../lib/utils/validations', () => {
+  return {
+    assertPayload: jest.fn(),
+    assertDefaultLocale: jest.fn().mockImplementationOnce(() => { throw new Error('ContentfulMltiError') })
```
Comment (axe312ger) [inline test/unit/index.test.js]: `ContentfulMultiError` - u missed the u

File: lib/parseOptions.js [hunk 8]
```diff
@@ -57,7 +57,7 @@ export default function parseOptions (params) {
 
   const proxySimpleExp = /.+:\d+/
   const proxyAuthExp = /.+:.+@.+:\d+/
-  if (options.proxy && !(proxySimpleExp.test(options.proxy) || proxyAuthExp.test(options.proxy))) {
+  if (typeof options.proxy === 'string' && options.proxy && !(proxySimpleExp.test(options.proxy) || proxyAuthExp.test(options.proxy))) {
```
Comment (axe312ger) [inline lib/parseOptions.js]: Can you duplicate this to https://github.com/contentful/contentful-export/blob/master/lib/parseOptions.js#L44 plz? kthxbye

File: lib/utils/validations.js (hunk 6 above)
Comment (Khaledgarbaya) [reply lib/utils/validations.js]: my idea is to map this key to the key in the JSON payload. I think it is fine like that

**Comment by axe312ger** [comment]
:shipit:
